### ISSUE TITLE
Real-time test discovery

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#2733]
ReSharper, NCrunch and now VSTest use static analysis to recognize tests as you type. Waiting for the compilation pipeline and then invoking NUnit on the final assembly brings such latency. I'm sympathetic to this need, driven more by user demand than anything else really.

Both of them mess up on things like `TestName` with `{m}` by showing bogus nodes with literal `{m}` in the name, and I can only imagine what will happen when we add `ArgDisplayNames`.

In order to avoid forcing them to be forever catching up to changes we make, I wish we could take responsibility of that static analysis by building a new, standalone assembly from this repository which returns tests as consistently as possible as if we had an actual assembly. Any PR that changes test discovery, such as SetArgDisplayNames, would be required to update the static analysis assembly to match.

Is this an idea worth pursuing in your opinions? Could we reach out and determine if that's something that any of these three would find helpful?

**Comment by rprouse** [comment]
I agree with you especially now that 15.6 is doing static analysis. They are using Roslyn to do the analysis and I've worked with the team to ensure they take into account as many variations as possible. I'm not sure we can take this over though. There are no hooks provided, so we would have to work with the VSTest team and there hasn't been interest on their side so far. I'll see what we can do. It will need to be as quick as the Rosyln analysis though.

**Comment by Barsonax** [comment]
Are there any plans on fixing this? This currently makes nunit unusable as it fails to discover tests or discovers tests that arent there.

**Comment by rprouse** [comment]
We are working with the Visual Studio team to improve their discovery, but it is partially out of our hands. It is Visual Studio doing the incorrect discovery. If you are experiencing serious problems, try turning off source based discovery in the settings.
